mls-rs-uniffi/src/lib.rs

```diff
@@ -0,0 +1,528 @@
+        let welcome_message = arc_unwrap_or_clone(welcome_message);
+        let (group, new_member_info) = self.inner.join_group(None, welcome_message.inner).await?;
+
+        let group = Arc::new(Group {
+            inner: Arc::new(Mutex::new(group)),
+        });
+        let group_info_extensions = Arc::new(new_member_info.group_info_extensions.into());
+        Ok(JoinInfo {
+            group,
+            group_info_extensions,
+        })
+    }
+}
+
+#[derive(Clone, Debug, uniffi::Object)]
+pub struct CommitOutput {
+    _inner: mls_rs::group::CommitOutput,
+}
+
+impl From<mls_rs::group::CommitOutput> for CommitOutput {
+    fn from(inner: mls_rs::group::CommitOutput) -> Self {
+        Self { _inner: inner }
+    }
+}
+
+#[derive(Clone, Debug, uniffi::Object)]
+pub struct SigningIdentity {
+    inner: identity::SigningIdentity,
+}
+
+impl From<identity::SigningIdentity> for SigningIdentity {
+    fn from(inner: identity::SigningIdentity) -> Self {
+        Self { inner }
+    }
+}
+
+/// An MLS end-to-end encrypted group.
+///
+/// The group is used to send and process incoming messages and to
+/// add/remove users.
+///
+/// See [`mls_rs::Group`] for details.
+#[derive(Clone, uniffi::Object)]
+pub struct Group {
+    inner: Arc<Mutex<mls_rs::Group<Config>>>,
+}
+
+#[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+impl Group {
+    #[cfg(not(mls_build_async))]
+    fn inner(&self) -> std::sync::MutexGuard<'_, mls_rs::Group<Config>> {
+        self.inner.lock().unwrap()
+    }
+
+    #[cfg(mls_build_async)]
+    async fn inner(&self) -> tokio::sync::MutexGuard<'_, mls_rs::Group<Config>> {
+        self.inner.lock().await
+    }
+
+    async fn index_to_identity(&self, index: u32) -> Result<identity::SigningIdentity, Error> {
+        let group = self.inner().await;
+        let member = group
+            .member_at_index(index)
+            .ok_or(MlsError::InvalidNodeIndex(index))?;
+        Ok(member.signing_identity)
+    }
+}
+
+/// Extract the basic credential identifier from a  from a key package.
+#[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+async fn signing_identity_to_identifier(
+    signing_identity: &identity::SigningIdentity,
+) -> Result<Vec<u8>, Error> {
+    let identifier = basic::BasicIdentityProvider::new()
+        .identity(signing_identity, &mls_rs::ExtensionList::new())
+        .await
+        .map_err(|err| err.into_any_error())?;
+    Ok(identifier)
+}
+
+#[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+#[uniffi::export]
+impl Group {
+    /// Perform a commit of received proposals (or an empty commit).
+    ///
+    /// TODO: ensure `path_required` is always set in
+    /// [`MlsRules::commit_options`](`mls_rs::MlsRules::commit_options`).
+    ///
+    /// Returns the resulting commit message. See
+    /// [`mls_rs::Group::commit`] for details.
+    pub async fn commit(&self) -> Result<Message, Error> {
```

Comment:
We probably want this to return CommitOutput as there are other properties in there you will require. Can you just add the function to get the commit message from CommitOutput for now?

Reply:
Let me start by returning the full `CommitOutput` here like I do elsewhere. I also added a few methods to `CommitOutput` now like @mulmarta suggested below.